sql/core/src/test/scala/org/apache/spark/sql/StringFunctionsSuite.scala

```diff
@@ -424,6 +424,14 @@ class StringFunctionsSuite extends QueryTest with SharedSparkSession {
       df.selectExpr("substring_index(a, '.', 2)"),
       Row("www.apache")
     )
+
+    val testTable = "test_substring_index"
+    withTable(testTable) {
+      sql(s"CREATE TABLE $testTable (num int) USING parquet")
+      sql(s"INSERT INTO $testTable VALUES (1), (2), (3), (NULL)")
```

Comment:
I'm fine to test with NULL, just to clarify: can we trigger this bug without NULL values?

Reply:
Oh, I see. I double-checked and it can also trigger this bug by correct numbers without null values. And I also updated the title and description of this PR.

Reply:
Furthermore, It seems that there are some deficiencies in our current testing mechanism related to built-in function with codegen, resulting in the inability to cover these scenarios. I don’t know if the codegen part of other functions can also be tested. If it is necessary to optimize the test coverage of this area, I am willing to do it with your support.

Reply:
A lot of expression tests simply use literal input, and we may miss some bugs in codegen which can only be exposed with column input. I think it will be the best if most expressions have unified implementation for interpreted and codegen, like `AesEncrypt`, the actual implementation is in Java. For other expressions, we need to manually test both literal input and column input.